query-planner-js/src/composedSchema/buildComposedSchema.ts

```diff
@@ -0,0 +1,240 @@
+      )
+    ) {
+      throw new GraphQLError(
+        `Unsupported core schema feature and/or version: ${feature}`,
+        schema.astNode!,
+      );
+    }
+  }
+
+  const joinName = 'join';
+
+  function getJoinDirective(name: string) {
+    const fullyQualifiedName = `${joinName}__${name}`;
+
+    const directive = schema.getDirective(fullyQualifiedName);
+    assert(
+      directive,
+      `Composed schema should define @${fullyQualifiedName} directive`,
+    );
+    return directive;
+  }
+
+  const ownerDirective = getJoinDirective('owner');
+  const typeDirective = getJoinDirective('type');
+  const fieldDirective = getJoinDirective('field');
+  const graphDirective = getJoinDirective('graph');
+
+  const graphEnumType = schema.getType(`${joinName}__Graph`);
+  assert(isEnumType(graphEnumType));
+
+  const graphMap: GraphMap = Object.create(null);
+
+  schema.extensions = {
+    ...schema.extensions,
+    federation: {
+      graphs: graphMap,
+    },
+  };
+
+  for (const graphValue of graphEnumType.getValues()) {
+    const name = graphValue.name;
+
+    const graphDirectiveArgs = getArgumentValuesForDirective(
+      graphDirective,
+      graphValue.astNode!,
+    );
+    assert(
+      graphDirectiveArgs,
+      `${graphEnumType.name} value ${name} in composed schema should have a @${graphDirective.name} directive`,
+    );
+
+    const serviceName: string = graphDirectiveArgs['name'];
+    const url: string = graphDirectiveArgs['url'];
+
+    graphMap[name] = {
+      serviceName,
+      url,
+    };
+  }
+
+  for (const type of Object.values(schema.getTypeMap())) {
+    if (isIntrospectionType(type)) continue;
+
+    // We currently only allow join spec directives on object types.
+    if (!isObjectType(type)) continue;
+
+    assert(
+      type.astNode,
+      `GraphQL type "${type.name}" should contain AST nodes`,
+    );
+
+    const ownerDirectiveArgs = getArgumentValuesForDirective(
+      ownerDirective,
+      type.astNode,
+    );
+
+    const typeMetadata: FederationTypeMetadata = ownerDirectiveArgs
+      ? {
+          serviceName: graphMap[ownerDirectiveArgs?.['graph']].serviceName,
+          keys: new MultiMap(),
+          isValueType: false,
+        }
+      : {
+          isValueType: true,
+        };
+
+    type.extensions = {
+      ...type.extensions,
+      federation: typeMetadata,
+    };
+
+    const typeDirectivesArgs = getArgumentValuesForRepeatableDirective(
+      typeDirective,
+      type.astNode,
+    );
+
+    assert(
+      !(typeMetadata.isValueType && typeDirectivesArgs.length >= 1),
+      `GraphQL type "${type.name}" cannot have a @${typeDirective.name} \
+directive without an $${ownerDirective.name} directive`,
+    );
+
+    for (const typeDirectiveArgs of typeDirectivesArgs) {
+      const serviceName = graphMap[typeDirectiveArgs['graph']].serviceName;
+
+      const keyFields = parseFieldSet(typeDirectiveArgs['key']);
+
+      typeMetadata.keys?.add(serviceName, keyFields);
```

Comment:
Separate question: Is it OK for a type to have `@join__type` declared multiple times for the same graph? Here it looks like you want it to be OK (use of MultiMap). I do see "It is an error to @join an object against the same subgraph multiple times." in the current spec draft but is that something we are changing? I guess it looks like there can be multiple from the printSupergraphSdl implementation, so I guess this is a spec issue.

Reply:
Yeah, that should be allowed according to the Federation model, but it turns out we have a composition validation rule against it (which was put in there because the query planner currently only uses the first key). So this is something we explicitly need to change both in the spec and in composition.